examples/nlp/machine_translation/create_tarred_dataset.py

```diff
@@ -0,0 +1,185 @@
+        cache_ids=False,
+        cache_data_per_node=False,
+        use_cache=False,
+    )
+    print('Batchifying ...')
+    dataset.batchify(encoder_tokenizer, decoder_tokenizer)
+    total_batches = len(dataset)
+    num_batches_per_tarfile = math.ceil(total_batches / args.num_tar_files_per_dataset_fragment)
+    cur_batches = 0
+    local_tar_file_ctr = 1
+    tar_file_ctr += 1
+    f_tar = tarfile.open(os.path.join(
+        args.out_dir,
+        'batches.tokens.%d.%d.tar' % (num_tokens, tar_file_ctr)
+    ), 'w')
+
+    for idx, (_, batch) in enumerate(dataset.batches.items()):
+        global_batch_ctr += 1
+        pickle.dump(batch, open(os.path.join(
+            args.out_dir,
+            'batch-%d.pkl' % (global_batch_ctr)
+        ), 'wb'))
+        f_tar.add(os.path.join(args.out_dir, 'batch-%d.pkl' % (global_batch_ctr)))
+        os.remove(os.path.join(args.out_dir, 'batch-%d.pkl' % (global_batch_ctr)))
+        cur_batches += 1
+        if (
+            cur_batches == num_batches_per_tarfile and
+            idx != len(dataset.batches) - 1
+        ):
+            f_tar.close()
+            cur_batches = 0
+            tar_file_ctr += 1
+            local_tar_file_ctr += 1
+            print('Creating local tar file %d' % (local_tar_file_ctr))
+            f_tar = tarfile.open(os.path.join(
+                args.out_dir,
+                'batches.tokens.%d.%d.tar' % (num_tokens, tar_file_ctr)
+            ), 'w')
+    f_tar.close()
+    return tar_file_ctr, global_batch_ctr
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='NMT dataset pre-processing')
+    parser.add_argument('--shared_tokenizer', action="store_true", help='Whether to share encoder/decoder tokenizers')
+    parser.add_argument('--clean', action="store_true", help='Whether to clean dataset based on length diff')
+    parser.add_argument('--bpe_dropout', type=float, default=0.1, help='Whether to share encoder/decoder tokenizers')
+    parser.add_argument('--src_fname', type=str, required=True, help='Path to the source file')
+    parser.add_argument('--tgt_fname', type=str, required=True, help='Path to the target file')
+    parser.add_argument('--out_dir', type=str, required=True, help='Path to store dataloader and tokenizer models')
+    parser.add_argument('--vocab_size', type=int, default=32000, help='Vocab size after BPE')
+    parser.add_argument('--max_seq_length', type=int, default=512, help='Max Sequence Length')
+    parser.add_argument('--min_seq_length', type=int, default=1, help='Min Sequence Length')
+    parser.add_argument(
+        '--tokens_in_batch', type=int, default=16000, help='# Tokens per batch per GPU'
+    )
+    parser.add_argument(
+        '--lines_per_dataset_fragment', type=int, default=1000000, help='# Tokens per dataset fragment'
+    )
+    parser.add_argument(
+        '--num_tar_files_per_dataset_fragment', type=int, default=4, help='Number of tarfiles'
+    )
+
```

Comment:
You cannot have uneven number of files inside each tar file, and the contents of each file should approximately be the same. Uneven data size inside tarfile will cause one or more workers to terminate before the rest, leaving DDP in a locked state. Look at https://github.com/NVIDIA/NeMo/blob/main/scripts/convert_to_tarred_audio_dataset.py#L127-L135 to see how ASR slices off small amounts of data to ensure that each tarfile has roughly the same amount of batches in it.

Reply:
Fixed! Thanks.